File: solutions/elixir/darts/3/lib/darts.ex
defmodule Darts do

  @rings [
    {:radius_lte, 1,  :inner},
    {:radius_lte, 5,  :middle},
    {:radius_lte, 10, :outer}
  ]

  def score({x, y}),      do: {x, y} |> zone() |> points()

  defp zone(coordinates), do: coordinates |> radius() |> ring()
  defp radius({x, y}),    do: :math.sqrt(x*x + y*y)

  defp ring(radius) do
    Enum.find_value(@rings, :outside, fn
      {_, limit, zone} when radius <= limit -> zone
      _ -> false
    end)
  end

  defp points(:inner),   do: 10
  defp points(:middle),  do: 5
  defp points(:outer),   do: 1
  defp points(:outside), do: 0

end

File: solutions/haskell/darts/3/src/Darts.hs
module Darts (score) where

type Position = (Float, Float)
data Zone = Inner | Middle | Outer | Outside

score :: Float -> Float -> Int
score x y = points . circle . radius $ (x, y)

radius :: Position -> Float
radius (x, y) = sqrt (x*x + y*y)

circle :: Float -> Zone
circle r
  | r <= 1    = Inner
  | r <= 5    = Middle
  | r <= 10   = Outer
  | otherwise = Outside

points :: Zone -> Int
points Inner   = 10
points Middle  = 5
points Outer   = 1
points Outside = 0
